### PR TITLE
testdrive: support DETAIL and HINT assertions

### LINKED ISSUE #15759: Add testdrive support for SQL error hints and details

### PR DESCRIPTION
**Pull request by guswynn**
Fixes #15759.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by benesch** [comment]
@chaas, you mind taking the review of this one, since you filed #15759?

**Review by philip-stoev — CHANGES_REQUESTED** [review]
Thank you. Please update doc/developer/testdrive.md with the expanded syntax. Thank you!

File: test/testdrive/structured-errors.td [hunk 1]
```diff
@@ -18,8 +18,16 @@
     PASSWORD SECRET pgpass
   )
 
-# TODO(guswynn): make testdrive support `startswith`, and hints and details
+# TODO(guswynn): make testdrive support `startswith`
 ! CREATE SOURCE "test_error"
   FROM POSTGRES CONNECTION pg (PUBLICATION 'mz_source')
   FOR ALL TABLES
 exact:failed to fetch publication information from PostgreSQL database
+detail:error connecting to server: Connection refused
```
Comment (philip-stoev) [inline test/testdrive/structured-errors.td:26]: As Buildkite reports the "Connection refused" part is OS-specific. Consider matching only on `error connecting to server`

**Review by chaas — APPROVED** [review]
Thanks for making this change!

File: src/testdrive/src/action/sql.rs [hunk 2]
```diff
@@ -288,6 +288,17 @@ impl fmt::Display for ErrorMatcher {
     }
 }
 
+impl ErrorMatcher {
+    fn fmt_with_type(&self, type_: &str) -> String {
```
Comment (chaas) [inline src/testdrive/src/action/sql.rs:292]: Rust beginner question - why did you underscore-suffix `type_`? Because it's a string literal?

**Comment by benesch** [comment]
"type" is a reserved keyword in Rust, so you have to come up with an
alternate spelling. (I'd say we typically use "ty" or "typ" elsewhere in
the codebase.)

File: test/testdrive/structured-errors.td (hunk 1 above)
Comment (guswynn) [reply test/testdrive/structured-errors.td:26]: Will do!

**Pull request merged**